src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Provides a method to read a binary log file (*.binlog) and replay all stored BuildEventArgs
+    /// by implementing IEventSource and raising corresponding events.
+    /// </summary>
+    public class BinaryLogReplayEventSource : EventArgsDispatcher
+    {
+        /// <summary>
+        /// Read the provided binary log file and raise corresponding events for each BuildEventArgs
+        /// </summary>
+        /// <param name="sourceFilePath">The full file path of the binary log file</param>
+        public void Replay(string sourceFilePath)
+        {
+            using (var stream = new FileStream(sourceFilePath, FileMode.Open))
+            {
+                var gzipStream = new GZipStream(stream, CompressionMode.Decompress, leaveOpen: true);
+                var binaryReader = new BinaryReader(gzipStream);
+
+                byte fileFormatVersion = binaryReader.ReadByte();
```

Comment:
Yeah I agree keep all the readers (and potentially writers) around and delegate to the right one. Current versions should be able to read all the old ones (when possible). In this case though we should throw if the version is greater than what we know about since there's no way it can read future versions.